src/Evolution/Systems/GrMhd/ValenciaDivClean/Subcell/TciOnFdGrid.cpp

```diff
@@ -9,29 +9,111 @@
+  evolution::dg::subcell::fd::reconstruct(
+      make_not_null(&get(dg_tilde_tau)), get(subcell_tilde_tau), dg_mesh,
+      subcell_mesh.extents(),
+      evolution::dg::subcell::fd::ReconstructionMethod::DimByDim);
+
+  // Note: `or` is short-circuiting, so if any of the first tests fail, we won't
+  // do the more expensive Persson TCI test.
   bool cell_is_troubled =
       vars_needed_fixing or
-      min(get(tilde_d)) <
+      min(get(dg_tilde_d)) <
           tci_options.minimum_rest_mass_density_times_lorentz_factor or
-      min(get(tilde_tau)) < tci_options.minimum_tilde_tau or
-      evolution::dg::subcell::persson_tci(tilde_d, dg_mesh, persson_exponent) or
-      evolution::dg::subcell::persson_tci(tilde_tau, dg_mesh, persson_exponent);
+      min(get(dg_tilde_tau)) < tci_options.minimum_tilde_tau or
+      evolution::dg::subcell::persson_tci(dg_tilde_d, dg_mesh,
+                                          persson_exponent) or
+      evolution::dg::subcell::persson_tci(dg_tilde_tau, dg_mesh,
+                                          persson_exponent);
+
+  if (cell_is_troubled) {
+    return {cell_is_troubled, rdmp_tci_data};
+  }
+
+  Scalar<DataVector> dg_mag_tilde_b{};
+  assign_data(make_not_null(&dg_mag_tilde_b), num_dg_pts);
+  evolution::dg::subcell::fd::reconstruct(
+      make_not_null(&get(dg_mag_tilde_b)), get(subcell_mag_tilde_b), dg_mesh,
+      subcell_mesh.extents(),
+      evolution::dg::subcell::fd::ReconstructionMethod::DimByDim);
+  // Add the reconstructed DG solution to the check. This is done so that we
+  // wouldn't violate RDMP if we switch back. However, we don't want to return
+  // max/mins from a bad DG solution.
+  using std::max;
+  using std::min;
+  evolution::dg::subcell::RdmpTciData rdmp_tci_data_for_check{};
+  rdmp_tci_data_for_check.max_variables_values = std::vector{
+      max(max(get(dg_tilde_d)), rdmp_tci_data.max_variables_values[0]),
+      max(max(get(dg_tilde_tau)), rdmp_tci_data.max_variables_values[1]),
+      max(max(get(dg_mag_tilde_b)), rdmp_tci_data.max_variables_values[2])};
+  rdmp_tci_data_for_check.min_variables_values = std::vector{
+      min(min(get(dg_tilde_d)), rdmp_tci_data.min_variables_values[0]),
+      min(min(get(dg_tilde_tau)), rdmp_tci_data.min_variables_values[1]),
+      min(min(get(dg_mag_tilde_b)), rdmp_tci_data.min_variables_values[2])};
+
+  cell_is_troubled |= evolution::dg::subcell::rdmp_tci(
+      rdmp_tci_data_for_check.max_variables_values,
+      rdmp_tci_data_for_check.min_variables_values,
+      past_rdmp_tci_data.max_variables_values,
+      past_rdmp_tci_data.min_variables_values, subcell_options.rdmp_delta0(),
+      subcell_options.rdmp_epsilon());
+
   if (tci_options.magnetic_field_cutoff.has_value() and not cell_is_troubled) {
-    const Scalar<DataVector> tilde_b_magnitude = magnitude(tilde_b);
-    cell_is_troubled = (max(get(tilde_b_magnitude)) >
+    cell_is_troubled = (max(get(dg_mag_tilde_b)) >
```

Comment:
It's only checking the Persson TCI if the magnetic field is large. This is because you lose spectral convergence as a solution goes to zero and so then you end up refining everywhere